### PR TITLE
Lindsay + Chantelle | Edges | OO Ride Share

### PR DESCRIPTION
**Pull request by BASIC-Belic**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer | Instructor Feedback
:------------- | :------------- | ---
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | we switched form DateTime to Time because subtraction was complicated and we needed to subtract time to `calculate_duration`
Describe any examples of composition that you encountered in this project, if any | We realized drivers can be users and inherit the same properties. Since it was in users, this helped us in populating things like name, id, etc in the driver instance. | This isn't composition, it's inheritance. We use the word `composition` to refer to has-a relationships, like "A trip has a driver" or "A passenger has a collection of trips"
Describe the relationship between `User` and `Driver`| User is a super class of driver. Driver has everything in user and more. However, not all drivers are Users in the passenger array. However Driver can be Driver class, User class, and still belong to the passenger array. 
Describe a _nominal test_ that you wrote for this assignment. | For `total_revenue` in Driver spec, we create two instances of a Trip, with made up costs. We did this calculation by hand and used test to verify that the program got the same result. 
Describe an _edge case test_ that you wrote for this assignment | For Driver spec we tested `initialize` driver to make sure it had a valid status. We changed the status to ":in_an_open_relationship," which would not work here. 
Describe a concept that you/your pair gained more clarity on as you worked on this assignment | The relationship between drivers and users, trips and driven trips. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | We need time to digest individually before collab. We made commit decisions together-- this was esp useful to not run into merge issues.

### CODE REVIEW

File: specs/trip_spec.rb [hunk 1]
```diff
@@ -9,36 +9,75 @@
...
+          end_time: end_time,
+          cost: 23.45,
+          rating: 3
+        }
+        @trip = RideShare::Trip.new(@trip_data)
+      end
+
+      it 'will calculate the duration in seconds' do
+        expect(@trip.calculate_duration).must_equal 25 * 60
+      end
```
Comment (droberts-sea) [inline specs/trip_spec.rb:82]: What will happen if there is no end time (i.e. trip is in progress?)

File: specs/user_spec.rb [hunk 2]
```diff
@@ -59,4 +59,65 @@
...
+        rating: 5)
+
+        @user.add_trip(trip1)
+        @user.add_trip(trip2)
+
+    end
+
+    it "returns the total amount that a user has spent on all of their rides" do
+
+    total_cost = @user.trips[0].cost + @user.trips[1].cost
```
Comment (droberts-sea) [inline specs/user_spec.rb:87]: These are a good start, but you're missing interesting cases for both these new methods:

- What happens if the user has no trips?
- What if the user has an incomplete trip?

File: specs/driver_spec.rb [hunk 3]
```diff
@@ -41,64 +55,106 @@
...
+              expect(@driver.average_rating).must_be_close_to (5.0 + 1.0) / 2.0, 0.01
+            end
+          end
+
+          describe "total_revenue" do
+            it 'will calculate revenue' do
+
+            start_time = '2015-05-20T12:14:00+00:00'
+            end_time = '2015-05-20T12:15:00+00:00'#
+            @driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ", status: :AVAILABLE)
```
Comment (droberts-sea) [inline specs/driver_spec.rb:125]: `@driver` here should be a local variable. You should only use an instance variable in your tests if it's coming from a `before` block.

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -41,64 +55,106 @@
...
+                expect(@driver.total_revenue).must_be_close_to 29.36
+              end
+          end
+
+          describe "net_expenditures" do
+
+        it 'will do something ' do
+              @dispatcher = RideShare::TripDispatcher.new(USER_TEST_FILE,TRIP_TEST_FILE,DRIVER_TEST_FILE)
+              start_time = '2015-05-20T12:14:00+00:00'
+              end_time = '2015-05-20T12:15:00+00:00'
```
Comment (droberts-sea) [inline specs/driver_spec.rb:146]: The tests for this method need some work. I would want to see the following cases examined:

- Driver has no trips, driven or ridden
- Driver has no driven trips but some ridden trips (should be negative number)
- Driver has driven trips but no ridden trips (should be positive number)
- Driver has made less driving than they've spent
- Driver has made more driving than they've spent
- The list of either driven or ridden trips includes an incomplete trip

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -29,44 +34,110 @@ def load_users(filename)
...
     end
 
+    def request_trip(user_id)
+
+      passenger = find_passenger(user_id)
+
+      this_driver = @drivers.find do |driver|
+        if driver.id != user_id
+          driver.status == :AVAILABLE
+        end
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:113]: It might make sense to move the logic for selecting a driver to a separate method.

File: lib/driver.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      end
+      return rating / @driven_trips.length
+
+    end
+
+    def total_revenue
+      total = 0.0
+      @driven_trips.each do |trip|
+        if trip.end_time != nil
+          total += 0.8 * (trip.cost - 1.65)
```
Comment (droberts-sea) [inline lib/driver.rb:44]: This code would be a little more readable if the two magic numbers (`0.8` and `1.65`) were stored in constants, maybe something like `DRIVER_FARE_PORTION` and `FEE_PER_TRIP`.

File: lib/trip_dispatcher.rb (hunk 5 above)
Comment (droberts-sea) [reply lib/trip_dispatcher.rb:113]: I'm also not 100% sure this code will do the right thing if `driver.id == user_id`. I think it would be more clear to say:

```ruby
this_driver = @drivers.find do |driver|
  driver.id != user_id && driver.status == :AVAILABLE
end
```

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -111,4 +106,23 @@
       expect(passenger.trips).must_include trip
     end
   end
+
+  describe "request trip method" do
+    it 'will populate a new trip with the necessary elements' do
+
+      dispatcher = RideShare::TripDispatcher.new(USER_TEST_FILE, TRIP_TEST_FILE, DRIVER_TEST_FILE)
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:113]: I would love to see some tests around driver availability here. For example, what happens if none of the drivers have status `:AVAILABLE`? What if the only driver who has that status is the passenger on this trip?

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes - good work
Answer comprehension questions | yes - see my feedback above
**Wave 1** |
Appropriate use of Ruby's `Date`| yes
`Trip` has a helper method to calculate duration| yes
`Passenger` has a method to calculate total cost of all trips| yes
Tests for wave 1 | see inline
**Wave 2** |
`Driver` inherits from `User` | yes
`Driver` has `add_driven_trip` method | yes
`Driver` has method to calculate average rating | yes
`Driver` has method to calculate net expenditures and it uses `super` | yes
`Driver` has a method to calculate total revenue | yes
Tests for wave 2 | see inline
**Wave 3** |
`TripDispatcher` has a new method to create trips | yes
creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Complex logic was correctly implemented | yes
Tests for `request_trip` | see inline
Methods from wave 1 and 2 handle incomplete trips | yes
Tests for wave 1 and 2 methods with incomplete trips | no
**Overall** |

Good job overall. Your product code is well-written, and for the most part your behavior seems to be correct.

The place where I see room for improvement is around testing. You've done a good job of building success test cases for each of your methods, but you are missing many tests for interesting edge cases. Since this is the first time you've written tests yourself, this is not a big concern, but it's definitely something to pay attention to on future projects.

In general I'm pretty happy with this submission, especially given how difficult this project ended up being. Keep up the hard work!
